bin/fetch-cards-pdf/main.go

```diff
@@ -6,135 +6,221 @@ import (
-	}
-	references := make(chan Ref)
-	go func() {
-		defer close(references)
-		doc.Find("carditem").EachWithBreak(func(_ int, s *goquery.Selection) bool {
-			references <- Ref{
-				ID:      s.AttrOr(":card", ""),
-				Faction: s.AttrOr("faction", ""),
-				Job:     s.AttrOr("job", ""),
-				Name:    s.AttrOr("title", ""),
-			}
-			return true
-		})
-	}()
+	// queue front cards for all refs
+	if !cfg.secondaryOnly {
+		go queueFirstCards(cfg.ppURL, jobQueue)
+	}
 
-	log.Debug("initializing imagick")
+	// queue front cards for special cases (colossal, character unit, dragoon)
+	go queueSpecialCaseCards(db, jobQueue)
 
-	imagick.Initialize()
-	defer imagick.Terminate()
+	// queue front cards for attachments (makeda & exalted court)
+	go queueAttachmentCards(db, jobQueue)
 
+	// process jobQueue
 	var wg sync.WaitGroup
-	for w := 0; w < workers; w++ {
+	for w := 0; w < cfg.workers; w++ {
 		wg.Add(1)
-		go func() {
+		go func(w int) {
 			defer wg.Done()
 			log.Debug("starting worker", "w", w)
 
-			mw := imagick.NewMagickWand()
-			defer mw.Destroy()
-
-			for ref := range references {
-				log := log.New("ref", ref.ID, "name", ref.Name)
-				log.Debug("retrieving card")
-
-				res, err := http.Get(fmt.Sprintf("%s/?card_items_to_pdf=$%s,1", baseURL, ref.ID))
-				if err != nil {
-					log.Error("retrieving card", "err", err)
+			for job := range jobQueue {
+				if err := downloadCard(cfg.ppURL, cfg.destDir, job.RefID, job.Index); err != nil {
+					log.Error("extract card", "err", err)
 					continue
 				}
+			}
+		}(w)
+	}
+	wg.Wait()
+}
 
-				raw, err := ioutil.ReadAll(res.Body)
-				if err != nil {
-					log.Error("reading card", "err", err)
-					continue
-				}
+func queueFirstCards(ppURL string, jobQueue chan DownloadJob) {
+	log.Debug("retrieving PP cards database")
+	res, err := http.Get(ppURL)
+	if err != nil {
+		log.Error("retrieving pp cards database", "err", err)
+		os.Exit(1)
+	}
 
-				mw.Clear()
+	log.Debug("parsing PP cards database")
+	doc, err := goquery.NewDocumentFromReader(res.Body)
+	if err != nil {
+		log.Error("parsing pp cards database", "err", err)
+		os.Exit(1)
+	}
 
-				err = mw.SetOption("density", "300")
-				if err != nil {
-					log.Error("setting wand options", "err", err, "option", "density", "value", "300")
-					return
-				}
+	log.Debug("looking for refs")
+	doc.Find("carditem").EachWithBreak(func(_ int, s *goquery.Selection) bool {
+		jobQueue <- DownloadJob{
+			RefID: s.AttrOr(":card", ""),
+			Index: 0,
+		}
+		return true
+	})
+}
 
-				err = mw.SetOption("quality", "100")
-				if err != nil {
-					log.Error("setting wand options", "err", err, "option", "quality", "value", "100")
-					return
-				}
+func queueSpecialCaseCards(db *sqlx.DB, jobQueue chan DownloadJob) {
+	log.Debug("looking for special cases")
 
-				err = mw.ReadImageBlob(raw)
-				if err != nil {
-					log.Error("parsing card", "err", err)
-					continue
-				}
+	stmt := `SELECT id, ppid, special FROM refs WHERE special != ""`
```

Comment:
Not sure why you put the statement in a separate variable here. It doesn't make the query more readable (and if you think it does you should do the same for the URL of the card database call L173).

Reply:
mostly habit, I like to have my queries isolated.